### PR TITLE
Improve bash completion experience.

### PR DESCRIPTION
**Pull request by denisfa**
Hey,

Made a lot of changes on the bash completion file to try to enhance the overall experience of using the command line tools. Although not a really high priority issue (since most should use `lxd` or similar), it would feel good if we could get a better completion.

I also realized that most distros have to create symlinks to use the bash completion file. The current out-of-the-box bash(>=4.1)  completion expects the command to have a completion that matches its name, i.e. if you try `lxc-create -B [TAB]`, it searches for `/usr/share/bash-completion/completions/lxc-create`, but this file does not exist. 
I tried my hand on a local solution, but if you feel that this is not time or even an issue, I will drop the commit and move forward with the rest. 

Most distros have these symlinks in their packages - debian, fedora, opensuse etc. Just a few - like my archlinux install - do not have a local solution. Not sure why this is a thing, but at least I learned a little bit of autotools.

As a last comment, I fixed some typos while working on the upgraded bash completion file. Hence the reason to add them here.

I would like to point out some extra stuff, but it is much easier to do on the code. See you there (hopefully).

### CODE REVIEW

File: config/bash/lxc.in [hunk 1]
```diff
@@ -1,111 +1,1031 @@
...
+            return
+            ;;
+        --rcfile | -f )
+            _filedir
+            return
+            ;;
+        --define | -s )
+            # @TODO: list values from source code.
+            # tag=value
+            # https://github.com/lxc/lxc/blob/stable-4.0/src/lxc/confile.c#L178
```
Comment (denisfa) [inline config/bash/lxc.in:530]: Ideally, this should be the output of a command from lxc. If it is not likely to change, though, we could just add the list manually like the templates and backing storages.

File: config/bash/lxc.in [hunk 2]
```diff
@@ -1,111 +1,1031 @@
...
-    _lxc_generic_t() {
-        local cur prev
+    if [[ $cur == -* ]]; then
+        COMPREPLY=( $( compgen -W '$( _parse_help "$1" )' -- "$cur" ) )
+        [[ ${COMPREPLY-} == *= ]] && compopt -o nospace
+        return
+    fi
+    _lxc_append_name
+}
+complete -F _lxc_cgroup lxc-cgroup
```
Comment (denisfa) [inline config/bash/lxc.in:257]: For `lxc-cgroup`, the last params are the `state-object` and its new value (optional). How feasible is something like:
```
__lxc_list_cgroup_state_object() {
    local userSlicePath="/sys/fs/cgroup/user.slice"
    if [[ -d "${userSlicePath}" ]] && [[ -f "${userSlicePath}/cgroup.controllers" ]]; then
        for cont in $(<"${userSlicePath}/cgroup.controllers"); do
            for entry in ${userSlicePath}/${cont}.*; do
                [[ ! -f "${entry}" ]] && continue
                local param="${entry##*/}"
                COMPREPLY+=( $( compgen -W "${param}" -- "$cur" ) )
            done
        done
    fi
}
```

to list the available values? Considering we don't want to use external tools.

**Comment by stgraber** [comment]
Oops, sorry, missed this one. Looking now.

File: config/bash/lxc.in (hunk 1 above)
Comment (stgraber) [reply config/bash/lxc.in:530]: That's a very long list and one that will pretty much always be partial as there are config keys that are dynamic (like the cgroup ones or environment keys). I personally tend to get annoyed when bash completion profiles try to return all the keys but then are missing a few.

My preference tends to be that either you can consistently list everything that's valid or you don't list anything.

**Comment by denisfa** [comment]
@stgraber Thank you for the feedback. Do not worry about missing this one, since it is a very low priority change. Just to be sure, was this merged intentionally? I made a mistake by adding commits directly into the master of my fork, so I forgot to rebase it on top of latest commit. Also I would like to try to work on solutions for the points listed on the code.

I had some cleanups to add plus changes depending on your feedback. Could I add them on a separate PR? In case this merge remains.

**Comment by stgraber** [comment]
It looked like an improvement and the commits were applying cleanly so I went ahead with it. Definitely feel free to open another PR to tweak things further!

**Comment by denisfa** [comment]
Thank you! Will do as soon as I can.

**Pull request merged**